connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -263,17 +267,20 @@ public boolean startConnector(
                 Plugins.compareAndSwapLoaders(savedLoader);
                 workerMetricsGroup.recordConnectorStartupFailure();
                 statusListener.onFailure(connName, t);
-                return false;
+                onConnectorStateChange.onCompletion(t, null);
+                return;
             }
+            workerConnector.transitionTo(initialState, onConnectorStateChange);
```

Comment:
This part still needs some work; it's in an inconsistent state because I modified `Worker::startConnector` to have no return value and instead communicate all success or failure of the connector startup through the callback, but haven't taken care of issues like possibly invoking the callback twice (once in this method, and once in the `WorkerConnector` instance), making sure to swap plugin classloaders at the right times, and preventing a possible race with the check to see if the connector already exists based on whether its name is present as a key in the `connectors` map.

Reply:
No race is possible; `connectors` is a `ConcurrentHashMap` and its `putIfAbsent` method is atomic.

We can just move this line (`workerConnector.transitionTo(initialState, onConnectorStateChange);`) back into the `try` block; it may throw an exception, but the callback we pass in will never be invoked until the `workerConnector` instance is submitted to the executor, so there should be no risk of it being invoked twice.